designs/2021-schema-object-rules/README.md

```diff
@@ -0,0 +1,291 @@
+
+There is an existing, third-party lint rule [eslint-plugin/prefer-object-rule](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/master/docs/rules/prefer-object-rule.md) to enforce and autofix that rules are written with the object-style.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The two proposals in this RFC should be implemented separately.
+
+### Detailed design for requiring schemas
+
+In `getRuleOptionsSchema()` in `eslint/lib/shared/config-validator.js` and in `@eslint/eslintrc/lib/shared/config-validator.js`, we will change the default schema from `null` to:
+
+```json
+{
+    "type": "array",
+    "minItems": "0",
+    "maxItems": "0"
+}
+```
+
+This will resulting in the following error message when running a test case that includes `options` for a rule without a schema:
+
+```pt
+Error: rule-tester:
+Configuration for rule "my-rule" is invalid:
+Value [{"foo-option":"foo-value"}] should NOT have more than 0 items.
+```
+
+or when passing options to a rule without a schema and executing linting with the rule:
+
+```pt
+Error: .eslintrc.js:
+Configuration for rule "my-rule" is invalid:
+Value [{"foo-option":"foo-value"}] should NOT have more than 0 items.
+```
+
+We'll then add tests in `eslint/tests/lib/shared/config-validator.js` and `@eslint/eslintrc/tests/lib/cascading-config-array-factory.js` that this error is correctly thrown.
+
+#### Opt-out
+
+In (hopefully) rare situations (e.g. bare-bones internal/private rules), we will allow users to opt-out from specifying a schema using `schema: false`. This clearly indicates that the user has chosen to forgo a schema. And a lint rule called `eslint-plugin/no-schema-opt-out` or `eslint-plugin/no-schema-false` could be created to discourage overuse/abuse of this opt-out.
+
+We believe this explicit opt-out is preferred over users resorting to hacky workarounds such as providing no-op/fake/minimal/incomplete schemas like `schema: {}` or `schema: { type: "array" }`. No-op schemas like these can confuse both humans (who might mistake it with `schema: []` which is not a no-op) and automated analysis/tooling built around schemas. So we will throw an error when encountering a few common versions of these no-op schemas:
+
+```pt
+`schema: {}` is a no-op. For rules with options, please fill in a complete schema. For rules without options, please omit `schema` or use `schema: []`.
+```
```

Comment:
If it causes a problem, I'm happy to revert this new exception. I figured this RFC would be a good chance to include it since it's very related, and you suggested it originally in the other thread. Definitely want to get the full TSC's opinion on it. I can fill in more details soon.

Reply:
In general, once an RFC is in final commenting, we shouldn’t be making any further additions, only adding clarifications.

I think it makes sense to bundle this change with the rest of the RFC but the additional details @mdjermanovic mentioned would be helpful.

Reply:
1. We can throw the exception in both core and rule tester. In core, we can throw it next to `ajv.compile` where the existing schema validation happens.
2. Honestly, I'm fine just checking for `schema: {}` since that's the most-common mistake and most-easy to mix up with `schema: []`. It should be easy to add more to the list later if there's an interest or need.
3. Yes I think we can log a warning in ESLint v8 when running tests for a rule with such a schema.

I used [eslint-ecosystem-analyzer](https://github.com/bmish/eslint-ecosystem-analyzer) and found that about ~7 of the top 1,000 plugins have an occurrence of `schema: {}`.

Reply:
I agreed, there is nothing the plugin users can do to avoid the error.  they could only disable the rule totally and report to the plugin author.

Reply:
> Although note that other schema validation happens in core already

Do we have validations of rule schemas in core? I think we've disabled Ajv validations and that we don't have any custom validations.

Reply:
My bad, I just confirmed that we don't validate schemas in core. We do `ajv.compile` in core but `ajv.validateSchema` only in rule tester.